### PR TITLE
[1725] Build rake task for rollover and RecruitmentCycle creation

### PR DESCRIPTION
**Pull request by avinhurry**
### Context

We are creating a service which encompasses the rollover code. It will be removed from MCB in a separate ticket. 

### Changes proposed in this pull request

- Add `RecruitmentCycleCreationService`, which takes the required fields as arguments. 
- Create `RolloverService` encompassing the logic previously in `rollover.rb`. You can rollover all providers or specific ones using a rake task.
- Refactor MCB rollover task, encapsulating logic in `RolloverService`
- Build rollover rake tasks

### Recruitment Cycle guidance to review

- Run ` bundle exec rake rollover:create_recruitment_cycle"[YYYY, "YYYY-MM-DD", "YYYY-MM-DD"]`. This is `year`, `application_start_date` and `application_end_date` respectively. 

- Open up the rails console and execute `RecruitmentCycle.find_by(year: YYYY)` to check the recruitment cycle has been successfully created. 

### Rollover guidance to review

- Ensure you have created a new `RecruitmentCycle`, as shown above because this is required.
- Ensure you have providers, locations and courses in your database.
- In Publish, change `can_edit_current_and_next_cycles` in `development.local.yml` to `true`
- Navigate to a provider `organisations/{provider_code}` check that you can click on `Current cycle` but not `Next cycle`
- Run `bundle exec rake rollover:providers"[{provider_code}]" `
- Go back to the publish UI and check that you can now access the next cycle. Check all sections have been correctly rolled over.
- You can rollover all providers, however there is currently an ongoing issue with email validation which will cause issues. See [this trello ticket](https://trello.com/c/dYKWlDZq/1416-fix-validation-on-provider-email).

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by avinhurry** [comment]
[Rollover - Write the copying code](https://trello.com/c/7K4xcwT4/1725-rollover-write-the-copying-code)

File: app/services/rollover_service.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+              copy_site_to_provider_service: Sites::CopyToProviderService.new,
+              logger: Logger.new(STDOUT),
+              )
+
+            counts = copy_provider_to_recruitment_cycle.execute(
+              provider: provider, new_recruitment_cycle: new_recruitment_cycle,
+              )
+          end
+        end
+        puts "provider #{counts[:providers] ? 'copied' : 'skipped'}, " \
```
Comment (AbigailMcP) [inline app/services/rollover_service.rb:47]: I don't think this ever prints `skipped`, even if we are skipping the provider. Based on the code in `Providers::CopyToRecruitmentCycleService` I think it should be: 
```suggestion
        puts "provider #{counts[:providers].zero? ? 'skipped' : 'copied'}, " \
```

File: lib/tasks/rollover.rake [hunk 2]
```diff
@@ -0,0 +1,11 @@
+namespace :rollover do
+  desc "Rollover providers, courses and locations to the next cycle"
+  task :providers, [:provider_code] => :environment do |_task, args|
```
Comment (AbigailMcP) [inline lib/tasks/rollover.rake:3]: In your service, you're doing `args.to_a.map(&:upcase)` which means you can pass in an array of provider codes. This makes is more obvious what you can pass in: 
```suggestion
  task :providers, [:provider_codes] => :environment do |_task, args|
```

**Review by AbigailMcP — COMMENTED** [review]
Well done! Couple of comments in line for tidying up the code, but I tested everything locally and is working well 💪

File: app/services/recruitment_cycle_creation_service.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+class RecruitmentCycleCreationService
```
[the remark this answers is not in the document]
Comment (avinhurry) [reply app/services/recruitment_cycle_creation_service.rb:1]: I agree, it's already in the folder name too.

File: lib/tasks/rollover.rake [hunk 4]
```diff
@@ -0,0 +1,12 @@
+namespace :rollover do
+  desc "Rollover providers, courses and locations to the next cycle"
+  task :providers, [:provider_codes] => :environment do |_task, args|
+    provider_codes = args[:provider_codes].split " "
```
Comment (AbigailMcP) [inline lib/tasks/rollover.rake:4]: `split` splits on space by default:

```suggestion
    provider_codes = args[:provider_codes].split
```

File: lib/tasks/rollover.rake [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+    RolloverService.call(provider_codes: provider_codes)
```
Comment (AbigailMcP) [inline lib/tasks/rollover.rake:5]: 👍 nice!

File: lib/tasks/rollover.rake (hunk 4 above)
Comment (AbigailMcP) [inline lib/tasks/rollover.rake:4]: How about when you run just `rake rollover:providers`? This will error at the moment.

**Review by AbigailMcP — COMMENTED** [review]
Just one thing to fix when calling the rake task with no arguments.

File: lib/tasks/rollover.rake (hunk 4 above)
Comment (avinhurry) [reply lib/tasks/rollover.rake]: Fixed

File: lib/tasks/rollover.rake [hunk 6]
```diff
@@ -0,0 +1,17 @@
+namespace :rollover do
+  desc "Rollover providers, courses and locations to the next cycle"
+  task :providers, [:provider_codes] => :environment do |_task, args|
+    provider_codes = args[:provider_codes]
+
+    if provider_codes.nil?
+      RolloverService.call(provider_codes: [provider_codes])
+    else
+      RolloverService.call(provider_codes: provider_codes.split)
+    end
```
Comment (AbigailMcP) [inline lib/tasks/rollover.rake:10]:
```suggestion
RolloverService.call(provider_codes: provider_codes&.split || [])
```

**Review by AbigailMcP — APPROVED** [review]

**Pull request merged**